lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -401,6 +443,58 @@ class BuildOpGroups : public calyx::FuncOpPartialLoweringPattern {
                         scf::ExecuteRegionOp executeRegionOp) const;
   LogicalResult buildOp(PatternRewriter &rewriter, CallOp callOp) const;
 
+  // Sets up the necessary state and resources for a `CmpIOp` in
+  // `buildLibraryBinaryPipeOp` if `cmpIOp` has sequential logic based on its
+  // operands.
+  void setupCmpIOp(PatternRewriter &rewriter, CmpIOp cmpIOp, Operation *group,
+                   calyx::RegisterOp &condReg, calyx::RegisterOp &resReg,
+                   bool &isSeqCondCheck) const {
+    isSeqCondCheck = (calyx::isSeqLibOpRes(cmpIOp.getLhs()) ||
+                      calyx::isSeqLibOpRes(cmpIOp.getRhs())) &&
+                     isa<scf::IfOp>(cmpIOp->getParentOp());
+    if (!isSeqCondCheck)
+      return;
+
+    StringRef opName = cmpIOp.getOperationName().split(".").second;
+    Type width = cmpIOp.getResult().getType();
+
+    condReg = createRegister(
+        cmpIOp.getLoc(), rewriter, getComponent(),
+        width.getIntOrFloatBitWidth(),
+        getState<ComponentLoweringState>().getUniqueName(opName));
+
+    for (auto *user : cmpIOp->getUsers()) {
+      if (auto ifOp = dyn_cast<scf::IfOp>(user))
+        getState<ComponentLoweringState>().setCondReg(ifOp, condReg);
+    }
+
+    assert(calyx::isSeqLibOpRes(cmpIOp.getLhs()) !=
+           calyx::isSeqLibOpRes(cmpIOp.getRhs()));
```

Comment:
I don't see why this would be illegal, so let's add a clear error message.

```suggestion
    assert(calyx::isSeqLibOpRes(cmpIOp.getLhs()) !=
           calyx::isSeqLibOpRes(cmpIOp.getRhs()) && "unexpected sequential operation on both sides; please open an issue");
```